[WIP]Make column iterator for SGMatrix

jonpsy opened the pull request:
Currently I'm getting this build error, I'm unsure if I should use pointer for m_mat. I'd like to know your thoughts ! :D. The unit-test case need some changes too, but first I want to clear the errors out.

https://pastebin.com/13tiQs3J

vigsterkr commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -258,6 +301,13 @@ template<class T> class SGMatrix : public SGReferencedData
 		const_iterator end() const noexcept { return const_iterator(matrix + (num_rows * num_cols)); }
 
+		/** Returns an iterator to the first element of the container. */
+		column_iterator begin_column() noexcept { return column_iterator(matrix); }
```
```suggestion
		column_iterator begin_column() noexcept { return column_iterator(*this); }
```

vigsterkr replied:
SGMatrix has no member called matrix...

vigsterkr commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,32 @@ TEST(SGMatrixTest,iterator)
+
+	//range-based loop should work as well
+	int idx;
+	std::for_each(mat.begin_column(),mat.end_column(),[&](SGVector<float64_t>* vec){ EXPECT_EQ(*vec,mat.get_column(idx++));});
```
```suggestion
	std::for_each(mat.begin_column(),mat.end_column(),[&idx](auto* vec){ EXPECT_EQ(*vec,mat.get_column(idx++));});
```

vigsterkr commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -258,6 +301,13 @@ template<class T> class SGMatrix : public SGReferencedData
+		column_iterator begin_column() noexcept { return column_iterator(matrix); }
+
+		/** Returns an iterator to the element following the last column of the container. */
+		column_iterator end_column() noexcept { return column_iterator(matrix + (num_rows * num_cols)); }
```
definitely not...

jonpsy replied:
Hmm 
```column_iterator end_column() noexcept{ return column_iterator(*this,num_cols); }``` ?

gf712 replied:
Yes

jonpsy replied:
.begin_column() and .end_column() are working perfectly fine !.  There's a small build error during for_each comparison. Need to fix that and we're done.

karlnapf commented:
It doesnt compile?

gf712 commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -41,6 +41,49 @@ template<class T> class SGMatrix : public SGReferencedData
+
+		private:
+			SGMatrix<T>& m_mat;
+			index_t m_idx;
```
Maybe call this m_column_idx, so it is clear what it is tracking

jonpsy replied:
Sure.

jonpsy replied:
How about ```m_col_idx``` ?

gf712 replied:
Sure!

gf712 commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,31 @@ TEST(SGMatrixTest,iterator)
+
+	//range-based loop should work as well
+	int idx;
+	std::for_each(mat.begin_column(),mat.end_column(),[&idx](auto* vec){ EXPECT_EQ(*vec,mat.get_column(idx++));});
```
Imo std::for_each is sufficient. The tests above are a bit convoluted..

jonpsy replied:
I'll fix that.

gf712 replied:
Thanks :)

jonpsy replied:
Have a check on them, see if it can be improved?

gf712 commented:
> It doesnt compile?

@NanuSai has to fix the end_column() class function

jonpsy commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,37 @@ TEST(SGMatrixTest,iterator)
+	constexpr index_t size=5;
+	SGMatrix<float64_t> mat(size, size);
+	linalg::range_fill(mat, 1.0);
+
+	SGMatrix<float64_t>::column_iterator col_itr(mat);
```
instead of this, should I use 
``` 
auto begin_col = mat.begin_column();
auto end_col = mat.end_column();
EXPECT_EQ(mat.num_cols, std::distance(begin_col, end_col));
EXPECT_TRUE(begin_col != end_col);
``` 

Sounds better to me (I compiled and tested it, it works )

jonpsy replied:
Okay, I implemented the same.

jonpsy commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,37 @@ TEST(SGMatrixTest,iterator)
+	EXPECT_EQ(mat.get_column(3),*new_itr);
+
+	//range-based loop should work as well
+	int idx;
+	std::for_each(mat.begin_column(),mat.end_column(),[&idx,mat](auto* vec){ EXPECT_EQ(*vec,mat.get_column(idx++));});
```
This doesn't work and throws this: https://pastebin.com/qArHsKcU

gf712 replied:
should be `std::for_each(mat.begin_column(),mat.end_column(),[&idx,mat](const auto& vec){ EXPECT_EQ(*vec,mat.get_column(idx++));}); `
You are passing a non-pointer type to a lambda expecting a pointer

jonpsy replied:
Okay using that gives https://pastebin.com/XVvURTpb. 
I think for_each deferences it for us, so we don't need to use ```*v``` again. It compiles succesfully when we use ```v```

gf712 replied:
You are dereferencing a non-pointer type without an operator* overload. Yes, you have to write it as `v`

jonpsy replied:
Right!.

gf712 commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,38 @@ TEST(SGMatrixTest,iterator)
+
+	//range-based loop should work as well
+	int idx=0;
+	std::for_each(mat.begin_column(),mat.end_column(),[&idx,mat](const auto& vec){ EXPECT_EQ(vec,mat.get_column(idx++));}); 
```
doesn't this cover all the test above?

jonpsy replied:
Hm, we do want to check if += , -= are working, right?

gf712 replied:
+= would be handled by for_each no? Anyway, I guess more tests can’t hurt!

jonpsy replied:
Haha yea, that's true. I meant += d btw. for_each increases by one only.And then there's also the deferencing one ```[]```

gf712 commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -41,6 +41,49 @@ template<class T> class SGMatrix : public SGReferencedData
+			bool operator < (const column_iterator &other) const { return m_col_idx < other.m_col_idx; }
+			bool operator > (const column_iterator &other) const { return m_col_idx > other.m_col_idx; }
+			bool operator <= (const column_iterator &other) const { return m_col_idx <= other.m_col_idx; }
+			bool operator >= (const column_iterator &other) const { return m_col_idx >= other.m_col_idx; }
```
Could you clang format this please? There is a mixture of styles here, such as spaces in operators and references on the variable vs on the type

jonpsy replied:
Sure!

jonpsy commented:
Woops ! Coming right up!

jonpsy commented:
I was thinking we should make mat.columns() so that we can have something like.. 
```for(auto& i: mat.columns ()) ```

Shall we go for that in this PR? I think since we're already knees deep in it, let's do it.

gf712 commented:
> I was thinking we should make mat.columns() so that we can have something like..
> `for(auto& i: mat.columns ()) `
> 
> Shall we go for that in this PR? I think since we're already knees deep in it, let's do it.

Yes, got for it. Do you know how to do it? You basically need columns to return a wrapper with begin and end class functions.

jonpsy commented:
I've made it, thanks for the guidance !.Unit tests are passing, maybe the description could be improved but other than that I think it's all good :!

LiuYuHui commented:
> 
> 
> I've made it, thanks for the guidance !.Unit tests are passing, maybe the description could be improved but other than that I think it's all good :!

it seems like that there are issues related to the SWIG in your code, maybe it can be fixed by just adding the  `#ifndef SWIG ` :)

jonpsy commented:
@LiuYuHui , thanks for the input. Correct me if I'm wrong, columns() method is already inside #ifndef,  I think you mean the implementation of "helper_struct"?.

gf712 commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -41,7 +41,244 @@ template<class T> class SGMatrix : public SGReferencedData
+				return !(*this == other);
+			}
+
+			value_type operator*()
```
Here you can return reference

jonpsy replied:
Correct me if I'm wrong but m_mat.get_column(m_col_idx) returns an SGVector<T>, so it's a local variable. We can't return a reference of local variable from functions right? (We can but that involves using value_type).

gf712 replied:
Ah yes, sorry

gf712 commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -41,7 +41,244 @@ template<class T> class SGMatrix : public SGReferencedData
+				return const_column_iterator(m_mat, m_col_idx - d);
+			}
+
+			value_type operator[](difference_type d) const
```
Same here

gf712 commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -41,7 +41,244 @@ template<class T> class SGMatrix : public SGReferencedData
+			auto end()
+			{
+				return m_mat.end_column();
+			}
```
Maybe also add the const versions?

jonpsy replied:
Oh right!, do I need another struct for this ?

jonpsy replied:
Hmm, maybe I can overload inside the struct. I'll try and let y'know ! :D

gf712 replied:
No, you can just do the same thing as in the iterators below. Just add begin/end const and add a columns const in SGMatrix

jonpsy commented:
Okay I've added support for const iterator. After much thoughts, I don't think there's any way to go about this without creating two helper_struct. I've also added a unit test to check it's working.  Would love to hear your feedback !

gf712 commented:
> Okay I've added support for const iterator. After much thoughts, I don't think there's any way to go about this without creating two helper_struct. I've also added a unit test to check it's working. Would love to hear your feedback !

technically you can, whether it is worth it is another question... @vigsterkr @karlnapf?
https://github.com/shogun-toolbox/shogun/blob/d80e9babce1f3f1da24e59e84b9025da77b2d483/src/shogun/util/container_iterators.h#L60-L67

gf712 commented:
LGTM. @vigsterkr should the iterator definitions be inside of SGMatrix? They take up so much space (in terms of LoC). If we keep them there we should have a single iterator that uses `std::conditional_t` to distinguish between const versions, reducing LoC to pretty much half..

jonpsy commented:
Okay that!!! Yes that would be amazing to have!!! I was thinking the same on code size, personally Id love to implement it

gf712 commented:
> Okay that!!! Yes that would be amazing to have!!! I was thinking the same on code size, personally Id love to implement it

OK, go ahead. You should just have to add an additional template parameter in the struct and the condition in the type definitions. You can do this with both the iterator and the helper struct

jonpsy commented:
If everything is good to go, now would be an excellent time to suggest a better name than "helper_struct".

vigsterkr commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -809,5 +554,254 @@ template<class T> class SGMatrix : public SGReferencedData
+			return m_mat.begin_column();
+		}
+
+		auto end()
```
`auto begin() const` and `auto end() const` is still missing...

vigsterkr commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -37,267 +37,19 @@ namespace shogun
-			SGMatrix<T>& m_mat;
-		};
+	private:
+		template<class ST>
```
```suggestion
		template<class MatrixType>
```

vigsterkr commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -37,267 +37,19 @@ namespace shogun
 {
 	friend class LinalgBackendEigen;
-
+	
```
noise

gf712 commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -527,5 +554,264 @@ template<class T> class SGMatrix : public SGReferencedData
+#ifndef SWIG
+	template<typename T>
+	template<class MatrixType>
+	struct SGMatrix<T>::helper_struct
```
I think just rename this to something more meaningful and it’s good to merge!

gf712 reviewed (changes requested):
Looks great, thanks!

gf712 commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,45 @@ TEST(SGMatrixTest,iterator)
+	index = 0;
+	for(const auto& i : const_mat.columns())
+		EXPECT_EQ(i,const_mat.get_column(index++));
+}
```
what about testing for modifying the value of the column vector?
```cpp
auto mat_copy = mat.clone();
for(auto& vec : mat_copy.columns())
	linalg::add(vec, 1, vec); //can't remember signature but basically vec += 1
for(const auto& [result, original] : zip_iterator(mat_copy.columns(), mat.columns()))
{
        auto reference = linalg::add(original, 1) // again can't remember signature but something like reference = original + 1 
	EXPECT_EQ(result, reference);
}
```

jonpsy replied:
It's this
```c++
1	//Modifying value of column vector
2	auto mat_copy = mat.clone();
3
4	for(const auto& vec : mat_copy.columns())
5		mat_copy.set_column(index,linalg::add(vec,vec));
6	
7	for(const auto& [result, original] : zip_iterator(mat_copy.columns(),mat.columns()))
8	{
9			auto reference = linalg::add(original,original)
10			EXPECT_EQ(result, reference);
11	}
```
However zip_iterator is showing at line 7. Here's the pastebin https://pastebin.com/DfjXGkuR
```
no instance of constructor "zip_iterator" matches the argument list -- argument types are: (shogun::SGMatrix<float64_t>::column_struct<shogun::SGMatrix<float64_t> &>, shogun::SGMatrix<float64_t>::column_struct<shogun::SGMatrix<float64_t> &>)
```

karlnapf commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -527,5 +553,264 @@ template<class T> class SGMatrix : public SGReferencedData
+		const SGMatrix<T>& m_mat;
+		difference_type m_col_idx;
+	};
+#ifndef SWIG
```
note that all of the iterator code should be hidden from swig, not just parts

jonpsy replied:
Ah cool, btw should I write "SWIG Should skip this part". I mean it's apparent anyway, thoughts?

karlnapf replied:
no need imo

karlnapf reviewed (commented):
nice thanks for this

gf712 commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,61 @@ TEST(SGMatrixTest,iterator)
+
+	auto idx1 = 0 ,idx2 =0;
+	while(index < std::distance(begin_col,end_col))
+		EXPECT_EQ(mat.get_column(idx1++),mat_copy.get_column(idx2++));
```
I think you missed the point of this test... You are now doing twice exactly the same and checking if the output is the same (which is a pretty basic assumption and I don't think we need to test for).
The test I suggested was to check whether updating the values in a loop was actually working when using the iterator. Please read the comment with the code snippet. What I meant was:
1. Mutate the vector from the iterator of a matrix
2. Compare in a separate loop so that we can be sure that mutating the vector returned by the iterator is actually changing the matrix itself and not a local copy. Hence you should either have a matrix with reference results or compute a reference value in a different way (e.g. do the same operation as you did in 1 with a local copy and compare that to the test matrix)

jonpsy committed:
Added
* column_iterator
* const_column_iterator
*  column_struct

jonpsy commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,55 @@ TEST(SGMatrixTest,iterator)
+	for (const auto& i : const_mat.columns())
+		EXPECT_EQ(i, const_mat.get_column(index++));
+
+	// Modifying value of column vector
+	auto mat_copy = mat.clone();
+
+	for (auto vec : mat.columns())
+		linalg::add_scalar(vec, 1.0);
+
+	auto range = std::distance(begin_col,end_col);
+	while (range-- > 0)
+		EXPECT_EQ(*mat.get_column(range), *mat_copy.get_column(range) + 1.0);
```
I believe this covers "mutating column vector"
* mat_copy gets a local copy of the original matrix
* We use a for loop then modify the vector returned by mat.columns() by adding 1.0 to it.
* Finally, we assert if the original matrix is changed or not (by comparing the difference in values)

gf712 replied:
OK, looks good. But could you write the while as a for loop, please? It is just increasing the code complexity.. Why not just `for (auto col_idx = 0; col_idx < mat.n_cols; col_idx++)`?

karlnapf replied:
or `for (auto col_idx in range(mat.n_cols))`

jonpsy committed:
for loop

gf712 commented:
Let's just wait for CI and it is good from my side. @vigsterkr @karlnapf what do you think?

gf712 commented on tests/unit/lib/SGMatrix_unittest.cc:
```diff
@@ -761,3 +761,54 @@ TEST(SGMatrixTest,iterator)
+	// Modifying value of column vector
+	auto mat_copy = mat.clone();
+
+	for (auto vec : mat.columns())
```
actually can you make this `auto&`?

jonpsy replied:
There's a problem, mat.columns() returns an rvalue.

gf712 replied:
ok, then just leave it like this

karlnapf replied:
we could add one version that returns a reference to the temporary SGVector? Would that make sense @gf712 ?

gf712 replied:
I don’t think that would work, because then that reference would be to a value that is stored somewhere. So the iterator would keep a “cached” vector, which could be fine. I mean the iterator is stateful anyway (we already store the index)

karlnapf commented on src/shogun/lib/SGMatrix.h:
```diff
@@ -41,7 +41,15 @@ template<class T> class SGMatrix : public SGReferencedData
 	public:
 		typedef RandomIterator<T> iterator;
 		typedef ConstRandomIterator<T> const_iterator;
+#ifndef SWIG	
```
@gf712 do we have to put all this code into SGMatrix.h? It bloats the file quite a bit? I'd prefer to have it somewhere else, where other iterators sit. But you or @vigsterkr might see this differently?

jonpsy replied:
Not a pro opinion but, I did try to have a go at pushing all this code to our ```<shogun/utils/iterators.h>```. This, however, creates a circular dependency issue because our iterator class  SGMatrix as an internal member and SGMatrix #includes iterator file.

karlnapf replied:
Yes, I feared that. Maybe the others have an idea

gf712 replied:
I mean it is doable, not sure if it will help. So you could have the declaration in iterator.h with forward declaration in SGMatrix and then implement the iterator in SGMatrix.cpp. This should work.

karlnapf commented:
What's the latest here? Would be cool to turn to get this in? :)

gf712 commented:
I think it’s ready to be merged!

karlnapf commented:
Thanks a lot for the patch @NanuSai 
Want to apply the new iterator to a few places in the codebase?

jonpsy commented:
> Thanks a lot for the patch @NanuSai
> Want to apply the new iterator to a few places in the codebase?

I'm game!
